pages/intelligence.md

```diff
@@ -0,0 +1,137 @@
+
+While investigating the above items, we will need to take into account
+the difference between usages and definitions. This will be pretty
+clear for most items (though aliasing and importing might muddy the
+waters somewhat), but atoms will be difficult, since there's no
+defintion to speak of, just a usage.
+
+# Caching
+> There are only two hard problems in computer science: Naming things, cache invalidation and off by one errors.
+
+Indexing the entire project, its dependencies, elixir and erlang when
+the language server starts is likely going to be too performance
+intensive to do (to wit, rewriting atoms during namespacing takes 15
+seconds on our moderately sized project), so we will need some sort of
+on-disk representation of the index.
+
+Such a representation will need to take the following into account:
+
+  * Branches can while the language server is running, invalidating large portions of the cache
+  * Project level search and replace can affect large numbers of files at once
+  * The versions of elixir and erlang can change via `asdf` whenever the user wants
+
+Basicallly, the project's assets can change at any time, and it's not
+guaranteed that the language server will be able to detect all of the
+changes, so the language server needs to treat this cache with some
+skepticism and constantly "check its work".
+
+## Cache layout
+A cache entry will have the following format:
+
+```elixir
+
+@type source :: Lexical.uri()
+@entry_type :: :otp | :elixir | :dependency | :project
+
+@type entry %{
+  type: entry_type(),
+  source: source(),
+  entry_version: String.t(),
+  added_at: DateTime.t(),
+  value: term()
+}
+```
+
+### Invalidation
+When lexical starts the first time, it will need to index the
+following entities in the following order:
+
+  1. The elixir release
+  2. All project dependencies
+  3. The project's source code and built artifacts
+
+The index needs to take into account which stage added it, and include an appropriate
+version when adding an entry. The versions are as follows:
+
+| Entry Type  | Entry Version |
+| ----------- | --------------|
+| :otp        | [This insanity](https://stackoverflow.com/questions/9560815/how-to-get-erlangs-release-version-number-from-a-shell/34326368#34326368)|
+| :elixir     | `System.version()`|
+| :dependency | `Application.spec(dep_atom)[:vsn]`|
+| :project    | `md5(file contents)` |
+
+Entries in the cache should be replaced and not updated, and we should
+have a mechanism to sweep old entries from the cache (for example, if
+a dependecy version is old and hasn't been updated in two weeks, we
+should remove it). This invalidation is going to be one of the tricky
+parts of the cache, and we should be aggressive with it, as
+re-indexing things is likely to be less bad than having an
+ever-growing cache on disk.  We need to think more about this and come
+up with something that strikes a balance between efficiency and
+on-disk storage.
```

Comment:
This paragraph confuses me. 

I guess the entire section is a little bit confusing. 

1. What is an `entity`?
2. Is an `entity` the same as an `entry`?
3. Is the cache going to hold `entities`/`entries` for the entire codebase or only some parts of the codebase?

Reply:
An entity in the cache is the thing actually being cached
An entry in the cache is an entity + some metadata that helps with eviction. 
The cache will cache everything that we index, which is the entire codebase + elixir + your dependencies.

Reply:
Got it.
So entities need to be indexed, all entities are indexed.
But when it comes caching, only some entities become and entry, based on... not sure.

Correct?